### PR TITLE
Implementation of more flexible hydration.

### PR DESCRIPTION
**Pull request by tsroka**
This allows defining links with multiple arguments source from object fields, field arguments or context

### CODE REVIEW

File: src/main/antlr/StitchingDSL.g4 [hunk 1]
```diff
@@ -26,9 +26,17 @@ serviceName: NAME;
 
 topLevelField: NAME;
 
+sourceObjectField : '$source' '.' name ;
```
Comment (tsroka) [inline src/main/antlr/StitchingDSL.g4]: I am happy with any naming changes or different way of expressing this in the grammar. Let me know if you have better ideas.

File: src/main/antlr/StitchingDSL.g4 [hunk 2]
```diff
@@ -26,9 +26,17 @@ serviceName: NAME;
...
+
+sourceFieldArgument : '$argument' '.' name ;
```
Comment (andimarek) [inline src/main/antlr/StitchingDSL.g4]: I think these are universal references. I would suggest ArgumentReference, ContextReference etc...

File: src/main/java/graphql/nadel/PrivateSchemaProvider.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package graphql.nadel;
+
+import com.atlassian.braid.SchemaNamespace;
+import graphql.schema.idl.TypeDefinitionRegistry;
+
+import java.util.Optional;
+
+public interface PrivateSchemaProvider {
```
Comment (johnliucn925) [inline src/main/java/graphql/nadel/PrivateSchemaProvider.java]: missing javadoc

File: src/main/java/graphql/nadel/dsl/RemoteArgumentSource.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static java.util.Collections.emptyList;
+
+public class RemoteArgumentSource extends AbstractNode<RemoteArgumentSource> {
```
Comment (johnliucn925) [inline src/main/java/graphql/nadel/dsl/RemoteArgumentSource.java]: missing javadoc

File: src/test/groovy/graphql/nadel/NadelTest.groovy [hunk 5]
```diff
@@ -301,6 +384,36 @@ class NadelTest extends Specification {
         return GraphQL.newGraphQL(graphQLSchema).build()
     }
 
+    /**
+     * Creates baz service with query that  returns Baz object with provided arguments
+     */
+    GraphQL bazService() {
+        def schema = """
```
Comment (johnliucn925) [inline src/test/groovy/graphql/nadel/NadelTest.groovy]: Could you change """ ->''' to align with other place you are doing in this PR ?

File: src/main/java/graphql/nadel/dsl/RemoteArgumentSource.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    public RemoteArgumentSource deepCopy() {
+        return null;
+    }
+
+    @Override
+    public TraversalControl accept(TraverserContext<Node> context, NodeVisitor visitor) {
+        return null;
+    }
+
+    public enum SourceType {OBJECT_FIELD, FIELD_ARGUMENT, CONTEXT}
```
Comment (andimarek) [inline src/main/java/graphql/nadel/dsl/RemoteArgumentSource.java]: please move that to the top of the class

File: build.gradle [hunk 7]
```diff
@@ -28,7 +28,7 @@ targetCompatibility = 1.8
 
 def slf4jVersion = '1.7.25'
 def graphqlJavaVersion = '10.0'
```
Comment (andimarek) [inline build.gradle]: I think it should be graphql-java 11.0, right?

File: src/main/antlr/StitchingDSL.g4 [hunk 8]
```diff
@@ -26,9 +26,17 @@ serviceName: NAME;
 
 topLevelField: NAME;
 
+sourceObjectReference : '$source' '.' name ;
+
+fieldArgumentReference : '$argument' '.' name ;
+
+contextArgumentReference : '$context' '.' name ;
+
+remoteArgumentSource : ( sourceObjectReference | fieldArgumentReference | contextArgumentReference ) ;
```
[the remark this answers is not in the document]
Comment (tsroka) [reply src/main/antlr/StitchingDSL.g4]: now I am unsure either :) I will remove them.

File: src/main/java/graphql/nadel/Nadel.java [hunk 9]
```diff
@@ -168,29 +182,36 @@ private Link createHydrationLink(SchemaNamespace schemaNamespace,
...
+
+        List<LinkArgument> linkArguments = hydration.getArguments().stream()
+                .map(argDef -> LinkArgument.newLinkArgument()
+                        .sourceName(argDef.getRemoteArgumentSource().getName())
+                        .argumentSource(LinkArgument.ArgumentSource.valueOf(argDef.getRemoteArgumentSource().getSourceType().toString()))
+                        .queryArgumentName(argDef.getName())
+                        .removeInputField(false)
+                        .nullable(true)
+                        .build())
+                .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (tsroka) [reply src/main/java/graphql/nadel/Nadel.java]: fixed!

**Pull request merged**